File: java-to-kotlin/src/main/kotlin/lec17/Lec17Main.java
package lec17;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class Lec17Main {

    public static void main(String[] args) {
        List<Fruit> fruits = Arrays.asList(
            new Fruit("사과", 1_000),
            new Fruit("사과", 1_200),
            new Fruit("사과", 1_200),
            new Fruit("사과", 1_500),
            new Fruit("바나나", 3_000),
            new Fruit("바나나", 3_200),
            new Fruit("바나나", 2_500),
            new Fruit("수박", 10_000)
        );

        List<Fruit> filtered1 = filterFruitsInterface(fruits, fruit -> fruit.getName().equals("사과"));
        for (Fruit fruit : filtered1) {
            System.out.println(fruit);
        }

        List<Fruit> filtered2 = filterFruitsStream(fruits, fruit -> fruit.getName().equals("사과"));
        for (Fruit fruit : filtered2) {
            System.out.println(fruit);
        }
    }

    private static List<Fruit> filterFruitsInterface(List<Fruit> fruits, Predicate<Fruit> fruitFilter) {
        List<Fruit> results = new ArrayList<>();
        for (Fruit fruit : fruits) {
            if (fruitFilter.test(fruit)) {
                results.add(fruit);
            }
        }
        return results;
    }

    private static List<Fruit> filterFruitsStream(List<Fruit> fruits, Predicate<Fruit> fruitFilter) {
        return fruits.stream()
                     .filter(fruitFilter)
                     .collect(Collectors.toList());
    }
}

File: java-to-kotlin/src/main/kotlin/lec17/Lec17Main.kt
package lec17

fun main() {
    val fruits = listOf(
        KoFruit("사과", 1_000),
        KoFruit("사과", 1_200),
        KoFruit("사과", 1_200),
        KoFruit("사과", 1_500),
        KoFruit("바나나", 3_000),
        KoFruit("바나나", 3_200),
        KoFruit("바나나", 2_500),
        KoFruit("수박", 10_000),
    )

    val isApple1: (KoFruit) -> Boolean = fun(fruit: KoFruit): Boolean {
        return fruit.name == "사과"
    }

    val isApple2: (KoFruit) -> Boolean = { fruit: KoFruit -> fruit.name == "사과" }

    println(isApple1(fruits[0]))
    println(isApple1.invoke(fruits[0]))
    println(isApple2(fruits[0]))
    println(isApple2.invoke(fruits[0]))

    val filtered1 = filterFruits(fruits, isApple1)
    val filtered2 = filterFruits(fruits, isApple2)
    val filtered3 = filterFruits(fruits, { fruit: KoFruit -> fruit.name == "사과" })
    val filtered4 = filterFruits(fruits) { fruit: KoFruit -> fruit.name == "사과" }
    val filtered5 = filterFruits(fruits) { fruit -> fruit.name == "사과" }
    val filtered6 = filterFruits(fruits) { it.name == "사과" }
    println("FIN")
}

private fun filterFruits(
    fruits: List<KoFruit>, filter: (KoFruit) -> Boolean
): List<KoFruit> {
    val results = mutableListOf<KoFruit>()
    for (fruit in fruits) {
        if (filter(fruit)) {
            results.add(fruit)
        }
    }
    return results
}